Mangler - reuse removed vars

boopathi opened the pull request.

boopathi commented on packages/babel-plugin-minify-mangle-names/src/index.js:
```diff
@@ -17,6 +17,32 @@ module.exports = ({ types: t }) => {
+
+    updateReference(scope, oldName, newName) {
+      if (!this.references.has(scope)) {
+        return;
```
this should throw ... new references aren't created when mangling.

boopathi commented:
<details>
<summary>We can do better for this - </summary>

```js
function Foo() {
  var a, b, c, d, e, f, g, h, i, j, k, l, m, n, o, p, q, r, s, t, u, v, w, x, y, z;
  var A, B, C, D, E, F, G, H, I, J, K, L, M, N, O, P, Q, R, S, T, U, V, W, X, Y, Z;
  var $, _;
  a, b, c, d, e, f, g, h, i, j, k, l, m, n, o, p, q, r, s, t, u, v, w, x, y, z;
  A, B, C, D, E, F, G, H, I, J, K, L, M, N, O, P, Q, R, S, T, U, V, W, X, Y, Z;
  $, _;
  function Foo() {
    var a, b, c, d, e, f, g, h, i, j, k, l, m, n, o, p, q, r, s, t, u, v, w, x, y, z;
    var A, B, C, D, E, F, G, H, I, J, K, L, M, N, O, P, Q, R, S, T, U, V, W, X, Y, Z;
    var $, _;
    a, b, c, d, e, f, g, h, i, j, k, l, m, n, o, p, q, r, s, t, u, v, w, x, y, z;
    A, B, C, D, E, F, G, H, I, J, K, L, M, N, O, P, Q, R, S, T, U, V, W, X, Y, Z;
    $, _;
    function Foo() {
      var a, b, c, d, e, f, g, h, i, j, k, l, m, n, o, p, q, r, s, t, u, v, w, x, y, z;
      var A, B, C, D, E, F, G, H, I, J, K, L, M, N, O, P, Q, R, S, T, U, V, W, X, Y, Z;
      var $, _;
      a, b, c, d, e, f, g, h, i, j, k, l, m, n, o, p, q, r, s, t, u, v, w, x, y, z;
      A, B, C, D, E, F, G, H, I, J, K, L, M, N, O, P, Q, R, S, T, U, V, W, X, Y, Z;
      $, _;
    }
  }
}
```

</details>

boopathi commented:
Applying this to jQuery - 

```
           raw     raw win gzip    gzip win parse time run    
babili     94.61kB 217%    32.71kB 158%     8ms        3952ms 
uglify     94.27kB 218%    32.78kB 158%     9ms        2427ms 
closure    94.23kB 218%    33.38kB 153%     9ms        10938ms
closure js 95.64kB 213%    33.78kB 150%     9ms        14664ms
```

boopathi commented on packages/babel-plugin-minify-mangle-names/src/index.js:
```diff
@@ -17,6 +17,36 @@ module.exports = ({ types: t }) => {
+      references.add(newName);
+    }
+
+    hasBinding(scope, name) {
```
TODO: remove

boopathi commented on packages/babel-plugin-minify-mangle-names/src/index.js:
```diff
@@ -101,11 +139,11 @@ module.exports = ({ types: t }) => {
           // Re-enable after enabling this feature
```
remove comments

boopathi commented on packages/babel-plugin-minify-mangle-names/src/index.js:
```diff
@@ -136,15 +174,16 @@ module.exports = ({ types: t }) => {
+              || mangler.hasReference(scope, next)
             );
 
             // TODO:
```
remove comments

boopathi commented on packages/babel-plugin-minify-mangle-names/src/index.js:
```diff
@@ -136,15 +174,16 @@ module.exports = ({ types: t }) => {
-            mangler.rename(scope, oldName, next);
+            resetNext();
+            mangler.rename(scope, binding, oldName, next);
+            mangler.updateReference(scope, oldName, next);
```
probably not necessary... have to double check

kangax commented:
@boopathi if you found no issues, and our tests pass, I would merge this (after removing comments and all that). Then I'll build this locally and try it on our codebase.

boopathi commented:
There are some bugs I'm noticing though .. Will update once I fix those ...

boopathi commented:
Updated and it looks like its working fine ... Except for one bug - when run with other plugins and when other plugins modify scope information, even though mangler is run as a separate pass, it disturbs this scope information and what we track in references becomes invalid for certain cases. 

<details>
<summary>Example:</summary>

```js
function foo() {
  path.traverse({
    VariableDeclaration: function VariableDeclaration(varPath) {
      // somehow these two affect the causing of the error
      if (!varPath.isVariableDeclaration({ kind: "var" })) return;
      if (!isSameFunctionScope(varPath, path)) return;

      var _iteratorNormalCompletion5 = true;
      var _didIteratorError5 = false;
      var _iteratorError5 = undefined;

      try {
        x();
      } catch (err) {
         // errors here in this scope
        _didIteratorError5 = true;
      }
    }
  });
}
```
</details>

boopathi commented:
weird that the tests pass on local, clearing cache on travis and re-running tests.

kangax commented:
> Except for one bug - when run with other plugins and when other plugins modify scope information, even though mangler is run as a separate pass, it disturbs this scope information and what we track in references becomes invalid for certain cases.

So we're kind of back to square 1? :( This is the exact issues I ran into when integrating mangler back in the days.

boopathi commented:
I fixed those things by creating them [on the fly](https://github.com/babel/babili/pull/243/commits/40a331cb504e648788ca0788b15dd1b9fb1cb946) ... and added some comments regarding the same ... Now it works

kangax commented:
Ok, I'll merge this in and try it out in a bit
